src/wasm/literal.cpp

```diff
@@ -33,36 +33,55 @@ Literal::Literal(const uint8_t init[16]) : type(Type::v128) {
   memcpy(&v128, init, 16);
 }
 
-Literal::Literal(const Literal& other) { *this = other; }
-
-Literal& Literal::operator=(const Literal& other) {
+Literal::Literal(const Literal& other) {
   type = other.type;
-  TODO_SINGLE_COMPOUND(type);
-  switch (type.getBasic()) {
-    case Type::i32:
-    case Type::f32:
-      i32 = other.i32;
-      break;
-    case Type::i64:
-    case Type::f64:
-      i64 = other.i64;
-      break;
-    case Type::v128:
-      memcpy(&v128, other.v128, 16);
-      break;
-    case Type::funcref:
-      func = other.func;
-      break;
-    case Type::exnref:
-      new (&exn) auto(std::make_unique<ExceptionPackage>(*other.exn));
-      break;
-    case Type::none:
-    case Type::nullref:
-      break;
-    case Type::externref:
-    case Type::unreachable:
-      WASM_UNREACHABLE("unexpected type");
+  if (type.isException()) {
+    if (other.exn.hasPackage) {
+      exn.hasPackage = true;
+      new (&exn.package) auto(
+        std::make_unique<ExceptionPackage>(*other.exn.package));
+    } else {
+      exn.hasPackage = false;
+    }
+  } else {
+    TODO_SINGLE_COMPOUND(type);
+    switch (type.getBasic()) {
+      case Type::i32:
+      case Type::f32:
+        i32 = other.i32;
+        break;
+      case Type::i64:
+      case Type::f64:
+        i64 = other.i64;
+        break;
+      case Type::v128:
+        memcpy(&v128, other.v128, 16);
+        break;
+      case Type::funcref:
+        func = other.func;
+        break;
+      case Type::none:
+        break;
+      case Type::externref:
+      case Type::anyref:
+      case Type::eqref:
+        break; // null
+      case Type::i31ref:
+      case Type::exnref:
+      case Type::unreachable:
+        WASM_UNREACHABLE("unexpected type");
+    }
   }
+}
+
+Literal& Literal::operator=(const Literal& other) {
+  // FIXME (dcode): Destructing the literal seems due, yet doing so corrupts
+  // `exn.package.values`, likely missing a check for equality or having
+  // something to do with passing around literals without `std::move`?
+  // Interestingly, code before the refactoring didn't destruct either.
+  // Affected test: wasm-shell test/spec/exception-handling.wast
+  // this->~Literal();
+  new (this) auto(other);
```

Comment:
This looked like a memory leak to me originally (assignment operator not destructing the enclosed unique_ptr), so I refactored the code to do what we ended up with when implementing `TypeInfo`, i.e. reusing constructor and destructor in the assignment operator. Turned out, though, that there is no equality operator to use to make this a nop where applicable, which might play a role, and that calling the destructor corrupts all but the first value in `exn.package.values` when running the test mentioned (in `br_on_exn_match_multiple_values`). The test `throw`s a package with `(i31.const 3) (f32.const 3.5)`, `catch`es it, assigns to a local and `rethrow`s, resulting in `(i31.const 3) (f32.const RANDOM_MEMORY)` when uncommenting the destructor call. Any ideas? So far thought about getting rid of the ´unique_ptr` in a union because it looks fragile to me.

Reply:
So, unfortunately, despite starring at this for the weekend, I wasn't able to pinpoint where stuff is going haywire. Changed around various things along the way, like made `Literal::type` immutable to guarantee that the payload is always properly recognized, scrapped the `unique_ptr` for manual `new/delete`, investigated `shared_ptr`, changed layouts and whatnot, but none of them helped :(